File: Airbridge/Airbridge/Scripts/AirbridgeUnity.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using UnityEngine;

/// <summary>
/// %Airbridge Unity SDK
/// </summary>
[SuppressMessage("ReSharper", "ClassNeverInstantiated.Global")]
[SuppressMessage("ReSharper", "CheckNamespace")]
public class Airbridge
{
    private static readonly Lazy<IAirbridgePlugin> Lazy = new Lazy<IAirbridgePlugin>(() =>
    {
        IAirbridgePlugin pluginInterface = new AirbridgeUnsupportedPlugin();
#if UNITY_ANDROID && !UNITY_EDITOR
            pluginInterface = new AirbridgeAndroidPlugin();
#elif UNITY_IOS && !UNITY_EDITOR 
            pluginInterface = new AirbridgeIOSPlugin();
#endif
            
        return pluginInterface;
    });
    
   private static IAirbridgePlugin AirbridgePlugin => Lazy.Value;
   
    #region Core

    /// <summary>
    /// Enables the SDK.
    /// </summary>
    public static void EnableSDK()
    {
        AirbridgePlugin.EnableSDK();
    }

    /// <summary>
    /// Disables the SDK.
    /// </summary>
    public static void DisableSDK()
    {
        AirbridgePlugin.DisableSDK();
    }

    /// <summary>
    /// Checks whether the SDK is currently enabled.
    /// </summary>
    /// <returns> `true` if the SDK is enabled, `false` otherwise.</returns>
    public static bool IsSDKEnabled()
    {
        return AirbridgePlugin.IsSDKEnabled();
    }
    
    #endregion

    #region Privacy

    /// <summary>
    /// Start collecting and transferring events.
    /// </summary>
    public static void StartTracking()
    {
        AirbridgePlugin.StartTracking();
    }

    /// <summary>
    /// Stop collecting and transferring events.
    /// </summary>
    public static void StopTracking()
    {
        AirbridgePlugin.StopTracking();
    }

    /// <summary>
    /// Checks whether the SDK is currently enabled for tracking.
    /// </summary>
    /// <returns> `true` if the SDK is enabled for tracking, `false` otherwise.</returns>
    public static bool IsTrackingEnabled()
    {
        return AirbridgePlugin.IsTrackingEnabled();
    }

    #endregion

    #region Deeplink

    /// <summary>
    /// Handles deeplink and deferred-deeplink.
    /// </summary>
    /// <param name="onDeeplinkReceived"> A callback listener that receives a delivered deeplink URL.</param>
    /// <example>
    /// **Example usage**
    /// @code
    /// Airbridge.SetOnDeeplinkReceived((string url)  =>
    /// {
    ///     /* Process deeplink data */
    /// });
    /// @endcode
    /// </example>
    public static void SetOnDeeplinkReceived(Action<string> onDeeplinkReceived)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetOnDeeplinkReceived(AirbridgeNullCheck.RequireNonNull(onDeeplinkReceived));
        });
    }

    #endregion

    #region Data Collection

    /// <summary>
    /// Sets the user ID.
    /// </summary>
    /// <param name="id"> The user ID.</param>
    public static void SetUserID(string id)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetUserID(AirbridgeNullCheck.RequireNonNull(id));
        });
    }

    /// <summary>
    /// Clears the user ID.
    /// </summary>
    public static void ClearUserID()
    {
        AirbridgePlugin.ClearUserID();
    }

    /// <summary>
    /// Sets the user email.
    /// </summary>
    /// <param name="email"> The user email.</param>
    public static void SetUserEmail(string email)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetUserEmail(AirbridgeNullCheck.RequireNonNull(email));
        });
    }

    /// <summary>
    /// Clears the user email.
    /// </summary>
    public static void ClearUserEmail()
    {
        AirbridgePlugin.ClearUserEmail();
    }

    /// <summary>
    /// Sets the user phone number.
    /// </summary>
    /// <param name="phone"> The user phone number.</param>
    public static void SetUserPhone(string phone)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetUserPhone(AirbridgeNullCheck.RequireNonNull(phone));
        });
    }

    /// <summary>
    /// Clears the user phone number.
    /// </summary>
    public static void ClearUserPhone()
    {
        AirbridgePlugin.ClearUserPhone();
    }

    /// <summary>
    /// Sets the key, value pair to the user attribute.
    /// </summary>
    /// <param name="key"> The key that uniquely identifies the user attribute.</param>
    /// <param name="value"> The value to set for the user attribute.</param>
    public static void SetUserAttribute(string key, object value)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetUserAttribute(AirbridgeNullCheck.RequireNonNull(key), AirbridgeNullCheck.RequireNonNull(value));
        });
    }

    /// <summary>
    /// Removes the user attribute with the given key.
    /// </summary>
    /// <param name="key"> The key that uniquely identifies the user attribute.</param>
    public static void RemoveUserAttribute(string key)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.RemoveUserAttribute(AirbridgeNullCheck.RequireNonNull(key));
        });
    }

    /// <summary>
    /// Clears all user attributes.
    /// </summary>
    public static void ClearUserAttributes()
    {
        AirbridgePlugin.ClearUserAttributes();
    }

    /// <summary>
    /// Sets the key, value pair to the user alias.
    /// </summary>
    /// <param name="key"> The key that uniquely identifies the user alias.</param>
    /// <param name="value"> The value to set for the user alias.</param>
    public static void SetUserAlias(string key, string value)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetUserAlias(AirbridgeNullCheck.RequireNonNull(key), AirbridgeNullCheck.RequireNonNull(value));
        });
    }

    /// <summary>
    /// Removes the user alias with the given key.
    /// </summary>
    /// <param name="key"> The key that uniquely identifies the user alias.</param>
    public static void RemoveUserAlias(string key)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.RemoveUserAlias(AirbridgeNullCheck.RequireNonNull(key));
        });
    }

    /// <summary>
    /// Clears all user aliases.
    /// </summary>
    public static void ClearUserAlias()
    {
        AirbridgePlugin.ClearUserAlias();
    }

    /// <summary>
    /// Clears all user information.
    /// </summary>
    public static void ClearUser()
    {
        AirbridgePlugin.ClearUser();
    }


    /// <summary>
    /// Sets the key, value pair to the device alias.
    /// </summary>
    /// <param name="key"> The key that uniquely identifies the device alias.</param>
    /// <param name="value"> The value to set for the device alias.</param>
    public static void SetDeviceAlias(string key, string value)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetDeviceAlias(AirbridgeNullCheck.RequireNonNull(key), AirbridgeNullCheck.RequireNonNull(value));
        });
    }

    /// <summary>
    /// Removes the device alias with the given key.
    /// </summary>
    /// <param name="key"> The key that uniquely identifies the device alias.</param>
    public static void RemoveDeviceAlias(string key)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.RemoveDeviceAlias(AirbridgeNullCheck.RequireNonNull(key));
        }); 
    }

    /// <summary>
    /// Clears all device aliases.
    /// </summary>
    public static void ClearDeviceAlias()
    {
        AirbridgePlugin.ClearDeviceAlias();
    }


    /// <summary>
    /// Registers a push notification token to track app uninstalls.
    /// </summary>
    /// <param name="token"> The push notification token, typically obtained from the device’s notification service provider,
    /// such as APNs for Apple devices or FCM registration token for Android devices.</param>
    public static void RegisterPushToken(string token)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.RegisterPushToken(AirbridgeNullCheck.RequireNonNull(token));
        });
    }

    #endregion

    #region Event

    /// <summary>
    /// Track events.
    /// </summary>
    /// <param name="category"> The category of the event.</param>
    /// <param name="semanticAttributes"> The semantic attributes of the event.</param>
    /// <param name="customAttributes"> The custom attributes of the event.</param>
    public static void TrackEvent(
        string category,
        Dictionary<string, object> semanticAttributes = null,
        Dictionary<string, object> customAttributes = null
    )
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.TrackEvent(AirbridgeNullCheck.RequireNonNull(category), semanticAttributes, customAttributes);
        });
    }

    #endregion

    #region Placement

    /// <summary>
    /// Queries the tracking link to the server to get the deeplink information and then moves according to that information if a deeplink is set up.
    /// It also adds a click event for the tracking link.
    /// </summary>
    /// <param name="trackingLink"> The URL to be queried for deeplink information and tracked for clicks.</param>
    /// <param name="onSuccess"> An optional callback action to be invoked if the click action is successful.</param>
    /// <param name="onFailure"> An optional callback action to be invoked if an error occurs during the click action.</param>
    /// <example>
    /// **Example usage**
    /// @code
    /// Airbridge.Click(
    ///     trackingLink: "https://abr.ge/~~~",
    ///     onSuccess: () => { /* Handle on success */ },
    ///     onFailure: (Exception exception) => { /* Handle on failure */ }
    /// );
    /// @endcode
    /// </example>
    public static void Click(string trackingLink, Action onSuccess = null, Action<Exception> onFailure = null)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.Click(AirbridgeNullCheck.RequireNonNull(trackingLink), onSuccess, onFailure);
        });
    }

    /// <summary>
    /// Adds an impression event for the tracking link.
    /// </summary>
    /// <param name="trackingLink"> The URL to be tracked for impressions.</param>
    /// <param name="onSuccess"> An optional callback action to be invoked if the impression action is successful.</param>
    /// <param name="onFailure"> An optional callback action to be invoked if an error occurs during the impression action.</param>
    /// <example>
    /// **Example usage**
    /// @code
    /// Airbridge.Impression(
    ///     trackingLink: "https://abr.ge/~~~",
    ///     onSuccess: () => { /* Handle on success */ },
    ///     onFailure: (Exception exception) => { /* Handle on failure */ }
    /// );
    /// @endcode
    /// </example>
    public static void Impression(string trackingLink, Action onSuccess = null, Action<Exception> onFailure = null)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.Impression(AirbridgeNullCheck.RequireNonNull(trackingLink), onSuccess, onFailure);
        });
    }

    #endregion

    #region Publication

    /// <summary>
    /// Request the Device UUID and get the response.
    /// </summary>
    /// <param name="onSuccess"> A callback action that takes a string parameter representing the Device UUID.</param>
    /// <param name="onFailure"> An optional callback action to be invoked if an error occurs.</param>
    /// <example>
    /// **Example usage**
    /// @code
    /// Airbridge.FetchDeviceUUID(
    ///     onSuccess: (string uuid) => { /* Process UUID */ },
    ///     onFailure: (Exception exception) => { /* Handle on failure */ }
    /// );
    /// @endcode
    /// </example>
    public static void FetchDeviceUUID(Action<string> onSuccess, Action<Exception> onFailure = null)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.FetchDeviceUUID(AirbridgeNullCheck.RequireNonNull(onSuccess), onFailure);
        });
    }

    /// <summary>
    /// Request the AirbridgeGeneratedUUID and get the response.
    /// </summary>
    /// <param name="onSuccess"> A callback action that takes a string parameter representing the %Airbridge Generated UUID.</param>
    /// <param name="onFailure"> An optional callback action to be invoked if an error occurs.</param>
    /// <example>
    /// **Example usage**
    /// @code
    /// Airbridge.FetchAirbridgeGeneratedUUID(
    ///     onSuccess: (string uuid) => { /* Process UUID */ },
    ///     onFailure: (Exception exception) => { /* Handle on failure */ }
    /// );
    /// @endcode
    /// </example>
    public static void FetchAirbridgeGeneratedUUID(Action<string> onSuccess, Action<Exception> onFailure = null)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.FetchAirbridgeGeneratedUUID(AirbridgeNullCheck.RequireNonNull(onSuccess), onFailure);
        });
    }

    /// <summary>
    /// Sets a listener for receiving attribution of install event.
    /// </summary>
    /// <param name="onAttributionReceived"> A callback listener that receives a Dictionary&lt;string, object&gt; parameter representing the attribution result.</param>
    /// <example>
    /// **Example usage**
    /// @code
    /// Airbridge.SetOnAttributionReceived((Dictionary&lt;string, object&gt; attributionResult) =>
    /// {
    ///     /* Process attribution data */
    /// });
    /// @endcode
    /// </example>
    public static void SetOnAttributionReceived(Action<Dictionary<string, object>> onAttributionReceived)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetOnAttributionReceived(AirbridgeNullCheck.RequireNonNull(onAttributionReceived));
        });
    }
    
    /// <summary>
    /// Creates a tracking-link using airbridge-server that move user to specific page of app and track click-event.
    /// </summary>
    /// <param name="channel"> The channel of tracking-link.</param>
    /// <param name="option"> The option to create tracking-link.</param>
    /// <param name="onSuccess"> A callback action that takes an AirbridgeTrackingLink parameter representing the tracking-link created.</param>
    /// <param name="onFailure"> An optional callback action to be invoked if an error occurs.</param>
    public static void CreateTrackingLink(
        string channel, 
        Dictionary<string, object> option, 
        Action<AirbridgeTrackingLink> onSuccess,
        Action<Exception> onFailure = null
    )
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.CreateTrackingLink(
                AirbridgeNullCheck.RequireNonNull(channel),
                AirbridgeNullCheck.RequireNonNull(option),
                AirbridgeNullCheck.RequireNonNull(onSuccess),
                onFailure
            );
        });
    }

    #endregion

    #region Hybrid

    /// <summary>
    /// Creates a script that initialize the web interface to implement web interface manually.
    /// </summary>
    /// <param name="webToken"> The token to initialize %Airbridge Web SDK.</param>
    /// <param name="postMessageScript"> The JavaScript code to post commands from web to app.
    /// (The parameter name used to send messages from web to app is `payload`.) </param>
    /// <returns> The web interface script.</returns>
    public static string CreateWebInterfaceScript(string webToken, string postMessageScript)
    {
        return AirbridgeNullCheck.CallMethodWithNullCheck(
            () => AirbridgePlugin.CreateWebInterfaceScript(AirbridgeNullCheck.RequireNonNull(webToken),
                AirbridgeNullCheck.RequireNonNull(postMessageScript)),
            defaultValue: ""
        );
    }

    /// <summary>
    /// Handles commands from the web interface to implement web interface manually.
    /// </summary>
    /// <param name="command"> The command to handle.</param>
    public static void HandleWebInterfaceCommand(string command)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.HandleWebInterfaceCommand(AirbridgeNullCheck.RequireNonNull(command));
        });
    }

    #endregion
    
    #region IAP
    
    /// <summary>
    /// Start collecting and transferring purchases events.
    /// </summary>
    public static void StartInAppPurchaseTracking()
    {
        AirbridgePlugin.StartInAppPurchaseTracking();
    }
    
    /// <summary>
    /// Stop collecting and transferring purchases events.
    /// </summary>
    public static void StopInAppPurchaseTracking()
    {
        AirbridgePlugin.StopInAppPurchaseTracking();
    }
    
    /// <summary>
    /// Checks whether the SDK is currently enabled for in-app purchase tracking.
    /// </summary>
    /// <returns> `true` if the SDK is enabled for in-app purchase tracking, `false` otherwise.</returns>
    public static bool IsInAppPurchaseTrackingEnabled()
    {
        return AirbridgePlugin.IsInAppPurchaseTrackingEnabled();
    }

    /// <summary>
    /// Sets a listener for receiving in-app purchase data.
    /// </summary>
    /// <param name="onReceived"> The listener to be set.</param>
    /// <example>
    /// **Example usage**
    /// @code
    /// Airbridge.SetOnInAppPurchaseReceived((ref AirbridgeInAppPurchase inAppPurchase) =>
    /// {
    /// #if UNITY_IOS
    ///     Debug.Log(DictionaryToString(inAppPurchase.Transaction));
    /// #elif UNITY_ANDROID
    ///     Debug.Log(DictionaryToString(inAppPurchase.Purchase));
    /// #endif
    ///     inAppPurchase.SetSemanticAttributes(new Dictionary&lt;string, object&gt;()
    ///     {
    ///         { "key", "value" }
    ///     });
    ///     inAppPurchase.SetCustomAttributes(new Dictionary&lt;string, object&gt;()
    ///     {
    ///         { "key", "value" }
    ///     });
    /// });
    /// @endcode
    /// </example>
    public static void SetOnInAppPurchaseReceived(OnAirbridgeInAppPurchaseReceiveListener onReceived)
    {
        AirbridgeNullCheck.CallMethodWithNullCheck(() =>
        {
            AirbridgePlugin.SetOnInAppPurchaseReceived(AirbridgeNullCheck.RequireNonNull(onReceived));
        });
    }

    #endregion

    /// <summary>
    /// Indicates whether notification was sent by %Airbridge to track uninstall of app.
    /// </summary>
    /// <param name="data"> The notification data to check.</param>
    /// <returns> `true` if notification was sent by %Airbridge, `false` otherwise.</returns>
    public static bool IsUninstallTrackingNotification(IDictionary<string, string> data)
    {
#if (UNITY_ANDROID || UNITY_IOS) && !UNITY_EDITOR
        return data.ContainsKey("airbridge-uninstall-tracking");
#else
        Debug.Log("Airbridge is not implemented this method on this platform.");
        return false;
#endif
    }
}

File: sdk-restricted/Plugins/Airbridge/Android/java/co/ab180/airbridge/unity/AirbridgeUnity.java
package co.ab180.airbridge.unity;

import static co.ab180.airbridge.unity.AirbridgeUtils.consumeIntent;
import static co.ab180.airbridge.unity.AirbridgeUtils.getMessage;
import static co.ab180.airbridge.unity.AirbridgeUtils.isIntentConsumed;
import static co.ab180.airbridge.unity.AirbridgeUtils.isNotNull;

import android.content.Intent;
import android.net.Uri;
import android.util.Log;

import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import co.ab180.airbridge.Airbridge;
import co.ab180.airbridge.AirbridgeLifecycleIntegration;

public class AirbridgeUnity {

    public static final String TAG = AirbridgeUnity.class.getSimpleName();

    private static String deeplink = null;
    private static AirbridgeCallback deeplinkCallback = null;

    private static String receivedAttributionResult = null;
    private static AirbridgeCallback attributionResultCallback = null;

    static AirbridgeCallbackWithReturn inAppPurchaseCallback = null;

    static AirbridgeLifecycleIntegration airbridgeLifecycleIntegration = null;

    public static void enableSDK() {
        try {
            Airbridge.enableSDK();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {enableSDK}", throwable);
        }
    }

    public static void disableSDK() {
        try {
            Airbridge.disableSDK();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {disableSDK}", throwable);
        }
    }

    public static boolean isSDKEnabled() {
        try {
            return Airbridge.isSDKEnabled();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {isSDKEnabled}", throwable);
        }
        return false;
    }

    public static void startInAppPurchaseTracking() {
        try {
            Airbridge.startInAppPurchaseTracking();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {startInAppPurchaseTracking}", throwable);
        }
    }

    public static void stopInAppPurchaseTracking() {
        try {
            Airbridge.stopInAppPurchaseTracking();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {stopInAppPurchaseTracking}", throwable);
        }
    }

    public static boolean isInAppPurchaseTrackingEnabled() {
        try {
            return Airbridge.isInAppPurchaseTrackingEnabled();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {isInAppPurchaseTrackingEnabled}", throwable);
        }
        return false;
    }
    
    public static void startTracking() {
        try {
            Airbridge.startTracking();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {startTracking}", throwable);
        }
    }

    public static void stopTracking() {
        try {
            Airbridge.stopTracking();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {stopTracking}", throwable);
        }
    }

    public static boolean isTrackingEnabled() {
        try {
            return Airbridge.isTrackingEnabled();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {isTrackingEnabled}", throwable);
        }
        return false;
    }

    public static void setUserID(@NotNull String id) {
        try {
            Airbridge.setUserID(id);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserID}", throwable);
        }
    }

    public static void clearUserID() {
        try {
            Airbridge.clearUserID();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {clearUserID}", throwable);
        }
    }

    public static void setUserEmail(@NotNull String email) {
        try {
            Airbridge.setUserEmail(email);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserEmail}", throwable);
        }
    }

    public static void clearUserEmail() {
        try {
            Airbridge.clearUserEmail();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {clearUserEmail}", throwable);
        }
    }

    public static void setUserPhone(@NotNull String phone) {
        try {
            Airbridge.setUserPhone(phone);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserPhone}", throwable);
        }
    }

    public static void clearUserPhone() {
        try {
            Airbridge.clearUserPhone();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {clearUserPhone}", throwable);
        }
    }

    public static void setUserAttribute(@NotNull String key, int value) {
        try {
            Airbridge.setUserAttribute(key, value);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserAttribute}", throwable);
        }
    }

    public static void setUserAttribute(@NotNull String key, long value) {
        try {
            Airbridge.setUserAttribute(key, value);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserAttribute}", throwable);
        }
    }

    public static void setUserAttribute(@NotNull String key, float value) {
        try {
            Airbridge.setUserAttribute(key, value);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserAttribute}", throwable);
        }
    }

    public static void setUserAttribute(@NotNull String key, double value) {
        try {
            Airbridge.setUserAttribute(key, value);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserAttribute}", throwable);
        }
    }

    public static void setUserAttribute(@NotNull String key, boolean value) {
        try {
            Airbridge.setUserAttribute(key, value);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserAttribute}", throwable);
        }
    }

    public static void setUserAttribute(@NotNull String key, @NotNull String value) {
        try {
            Airbridge.setUserAttribute(key, value);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserAttribute}", throwable);
        }
    }

    public static void removeUserAttribute(@NotNull String key) {
        try {
            Airbridge.removeUserAttribute(key);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {removeUserAttribute}", throwable);
        }
    }

    public static void clearUserAttributes() {
        try {
            Airbridge.clearUserAttributes();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {clearUserAttributes}", throwable);
        }
    }

    public static void setUserAlias(@NotNull String key, @NotNull String value) {
        try {
            Airbridge.setUserAlias(key, value);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setUserAlias}", throwable);
        }
    }

    public static void removeUserAlias(@NotNull String key) {
        try {
            Airbridge.removeUserAlias(key);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {removeUserAlias}", throwable);
        }
    }

    public static void clearUserAlias() {
        try {
            Airbridge.clearUserAlias();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {clearUserAlias}", throwable);
        }
    }

    public static void clearUser() {
        try {
            Airbridge.clearUser();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {clearUser}", throwable);
        }
    }

    public static void setDeviceAlias(@NotNull String key, @NotNull String value) {
        try {
            Airbridge.setDeviceAlias(key, value);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {setDeviceAlias}", throwable);
        }
    }

    public static void removeDeviceAlias(@NotNull String key) {
        try {
            Airbridge.removeDeviceAlias(key);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {removeDeviceAlias}", throwable);
        }
    }

    public static void clearDeviceAlias() {
        try {
            Airbridge.clearDeviceAlias();
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {clearDeviceAlias}", throwable);
        }
    }

    public static void registerPushToken(@NotNull String token) {
        try {
            Airbridge.registerPushToken(token);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {registerPushToken}", throwable);
        }
    }

    public static void trackEvent(
            @NotNull String category,
            String semanticAttributesJsonString,
            String customAttributesJsonString
    ) {
        try {
            Map<String, Object> semanticAttributes = AirbridgeJsonParser.from(semanticAttributesJsonString);
            Map<String, Object> customAttributes = AirbridgeJsonParser.from(customAttributesJsonString);
            Airbridge.trackEvent(category, semanticAttributes, customAttributes);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {trackEvent}", throwable);
        }
    }

    public static void click(@NotNull String trackingLink, AirbridgeCallback onSuccess, AirbridgeCallback onFailure) {
        try {
            Airbridge.click(
                    Uri.parse(trackingLink),
                    unit -> {
                        if (onSuccess != null) onSuccess.Invoke("");
                    },
                    throwable -> {
                        if (onFailure != null) onFailure.Invoke(getMessage(throwable));
                    }
            );
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {click}", throwable);
        }
    }

    public static void impression(@NotNull String trackingLink, AirbridgeCallback onSuccess, AirbridgeCallback onFailure) {
        try {
            Airbridge.impression(
                    Uri.parse(trackingLink),
                    unit -> {
                        if (onSuccess != null) onSuccess.Invoke("");
                    },
                    throwable -> {
                        if (onFailure != null) onFailure.Invoke(getMessage(throwable));
                    }
            );
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {impression}", throwable);
        }
    }

    public static void fetchAirbridgeGeneratedUUID(@NotNull AirbridgeCallback onSuccess, AirbridgeCallback onFailure) {
        try {
            Airbridge.fetchAirbridgeGeneratedUUID(
                    onSuccess::Invoke,
                    throwable -> {
                        if (onFailure != null) onFailure.Invoke(getMessage(throwable));
                    }
            );
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {fetchAirbridgeGeneratedUUID}", throwable);
        }
    }

    public static void fetchDeviceUUID(@NotNull AirbridgeCallback onSuccess, AirbridgeCallback onFailure) {
        try {
            Airbridge.fetchDeviceUUID(
                    onSuccess::Invoke,
                    throwable -> {
                        if (onFailure != null) onFailure.Invoke(getMessage(throwable));
                    }
            );
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {fetchDeviceUUID}", throwable);
        }
    }

    public static String createWebInterfaceScript(@NotNull String webToken, @NotNull String postMessageScript) {
        try {
            return Airbridge.createWebInterfaceScript(webToken, postMessageScript);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {createWebInterfaceScript}", throwable);
        }
        return "";
    }

    public static void handleWebInterfaceCommand(@NotNull String command) {
        try {
            Airbridge.handleWebInterfaceCommand(command);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {handleWebInterfaceCommand}", throwable);
        }
    }

    public static void createTrackingLink(
            @NotNull String channel,
            String optionJsonString,
            @NotNull AirbridgeCallback onSuccess,
            AirbridgeCallback onFailure
    ) {
        try {
            Map<String, Object> option = AirbridgeJsonParser.from(optionJsonString);
            if (option == null) {
                option = new HashMap<>();
            }
            Airbridge.createTrackingLink(channel, option,
                    trackingLink -> {
                        try {
                            JSONObject result = new JSONObject();
                            result.put("shortURL", trackingLink.getShortURL().toString());
                            result.put("qrcodeURL", trackingLink.getQrcodeURL().toString());
                            onSuccess.Invoke(result.toString());
                        } catch (Throwable throwable) {
                            Log.e(TAG, "Error occurs while parsing tracking-link data to json string", throwable);
                        }
                    },
                    throwable -> {
                        if (onFailure != null) onFailure.Invoke(getMessage(throwable));
                    }
            );
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {createTrackingLink}", throwable);
        }
    }
    
    public static void setOnInAppPurchaseReceived(AirbridgeCallbackWithReturn onInAppPurchaseReceived) {
        if (inAppPurchaseCallback == null) {
            inAppPurchaseCallback = onInAppPurchaseReceived;
        } else {
            Log.w(TAG, "Already called {setOnInAppPurchaseReceived}");
        }
    }

    /* ========================== Handle Deeplink ========================== */

    public static void handleDeeplink(@NotNull AirbridgeCallback onDeeplinkReceived) {

        // Only the first callback is valid
        if (deeplinkCallback != null) return;
        deeplinkCallback = onDeeplinkReceived;

        // Success
        if (isNotNull(deeplink)) {
            handleDeeplinkOnSuccess(deeplink);
        }
        // Check deferred deeplink
        else {
            processHandleDeferredDeeplink();
        }
    }

    @SuppressWarnings("WeakerAccess")
    public static void processHandleDeeplink(Intent intent) {
        try {
            boolean handled = Airbridge.handleDeeplink(intent, /* NotNull */ uri -> {
                if (isNotNull(deeplinkCallback)) {
                    handleDeeplinkOnSuccess(uri.toString());
                } else {
                    deeplink = uri.toString();
                }
            });

            if (handled || AirbridgeSettings.isHandleAirbridgeDeeplinkOnly) return;
            if (isIntentConsumed(intent)) {
                Log.d(TAG, "Intent was already consumed on {handleDeeplink}");
                return;
            }
            consumeIntent(intent);
            
            Uri data = intent.getData();
            if (data == null) return;

            if (isNotNull(deeplinkCallback)) {
                handleDeeplinkOnSuccess(data.toString());
            } else {
                deeplink = data.toString();
            }
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {handleDeeplink}", throwable);
        }
    }

    private static void processHandleDeferredDeeplink() {
        try {
            Airbridge.handleDeferredDeeplink(/* Nullable */ uri -> {
                if (isNotNull(uri) && isNotNull(deeplinkCallback)) {
                    handleDeeplinkOnSuccess(uri.toString());
                }
            });
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while calling {handleDeferredDeeplink}", throwable);
        }
    }

    private static void handleDeeplinkOnSuccess(String uri) {
        deeplinkCallback.Invoke(uri);
        deeplink = null;
    }

    /* ========================== Attribution Result Callback ========================== */

    public static void setOnAttributionReceived(@NotNull AirbridgeCallback onAttributionReceived) {

        // Only the first callback is valid
        if (attributionResultCallback != null) return;
        attributionResultCallback = onAttributionReceived;

        unitySendAttributionResult();
    }

    public static void processSetOnAttributionReceived(@NotNull Map<String, String> result) {
        try {
            receivedAttributionResult = AirbridgeJsonParser.to(result);
        } catch (Throwable throwable) {
            Log.e(TAG, "Error occurs while parsing attribution data to json string", throwable);
        }
        unitySendAttributionResult();
    }

    private static void unitySendAttributionResult() {
        if (isNotNull(attributionResultCallback) && isNotNull(receivedAttributionResult)) {
            attributionResultCallback.Invoke(receivedAttributionResult);
            receivedAttributionResult = null;
        }
    }

    /* ================================================================================= */

    public static void setLifecycleIntegration(AirbridgeLifecycleIntegration lifecycleIntegration) {
        if (airbridgeLifecycleIntegration != null) return;
        airbridgeLifecycleIntegration = lifecycleIntegration;
    }
}